File: Manuel/개인 활동/Class 3/최소 힙/Without STL/1927_No_STL.py
import sys

class MinHeap:
    def __init__(self):
        self.heap = []
    
    def _parent(self, i):
        return(i - 1) // 2
    def _left(self, i):
        return 2 * i + 1
    def _right(self, i):
        return 2 * i + 2
    def _heapify_up(self, i):
        while i > 0 and self.heap[self._parent(i)] > self.heap[i]:
            self.heap[self._parent(i)], self.heap[i] = self.heap[i], self.heap[self._parent(i)]
            i = self._parent(i)
    def _heapify_down(self, i):
        smallest = i
        l = self._left(i)
        r = self._right(i)
        
        if l < len(self.heap) and self.heap[l] < self.heap[smallest]:
            smallest = l
        if r < len(self.heap) and self.heap[r] < self.heap[smallest]:
            smallest = r
        if smallest != i:
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self._heapify_down(smallest)
    
    def push(self, val):
        self.heap.append(val)
        self._heapify_up(len(self.heap) - 1)
    def top(self):
        if not self.heap:
            return 0
        return self.heap[0]
    def pop(self):
        if not self.heap:
            return
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        self.heap.pop()
        self._heapify_down(0)
    def empty(self):
        return len(self.heap) == 0
    
def main():
    input = sys.stdin.readline

    N = int(input())
    min_heap = MinHeap()

    for _ in range(N):
        x = int(input())

        if x == 0:
            if min_heap.empty():
                print(0)
            else:
                print(min_heap.top())
                min_heap.pop()
        else:
            min_heap.push(x)

main()

File: Manuel/개인 활동/Class 3/최소 힙/Without STL/1927_No_STL.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define endl "\n"

using namespace std;

class MinHeap
{
private:
    vector<int> heap;
    int parent(int i)
    {
        return (i - 1) / 2;
    }
    int left(int i)
    {
        return 2 * i + 1;
    }
    int right(int i)
    {
        return 2 * i + 2;
    }

    void heapifyUp(int i)
    {
        while(i > 0 && heap[parent(i)] > heap[i])
        {
            swap(heap[parent(i)], heap[i]);
            i = parent(i);
        }
    }
    void heapifyDown(int i)
    {
        int smallest = i;
        int l = left(i);
        int r = right(i);

        if(l < heap.size() && heap[l] < heap[smallest])
        {
            smallest = l;
        }
        if(r < heap.size() && heap[r] < heap[smallest])
        {
            smallest = r;
        }
        if(smallest != i)
        {
            swap(heap[i], heap[smallest]);
            heapifyDown(smallest);
        }
    }
public:
    void push(int val)
    {
        heap.push_back(val);
        heapifyUp(heap.size() - 1);
    }
    int top()
    {
        if(heap.empty())
        {
            return 0;
        }
        return heap.front();
    }
    void pop()
    {
        if(heap.empty())
        {
            return;
        }
        swap(heap.front(), heap.back());
        heap.pop_back();
        heapifyDown(0);
    }
    bool empty()
    {
        return heap.empty();
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;

    MinHeap min_heap;

    for(int i = 0; i < N; i++)
    {
        int x;
        cin >> x;

        if(x == 0)
        {
            if(min_heap.empty())
            {
                cout << 0 << endl;
            }
            else
            {
                cout << min_heap.top() << endl;
                min_heap.pop();
            }
        }
        else
        {
            min_heap.push(x);
        }
    }
    return 0;
}